django/utils/numberformat.py

```diff
@@ -18,46 +17,93 @@ def format(number, decimal_sep, decimal_pos=None, grouping=0, thousand_sep='',
+    elif use_l10n is False:
+        grouping = 0
+    elif use_l10n is None and not settings.USE_L10N:
+        grouping = 0
+
     # Make the common case fast
-    if isinstance(number, int) and not use_grouping and not decimal_pos:
-        return mark_safe(number)
-    # sign
-    sign = ''
-    # Treat potentially very large/small floats as Decimals.
-    if isinstance(number, float) and 'e' in str(number).lower():
-        number = Decimal(str(number))
+    if isinstance(number, int) and (grouping == 0 or grouping == 3):
+        return _format_int(number, decimal_sep, decimal_pos, thousand_sep, grouping)
+
+    # Handle Decimals
     if isinstance(number, Decimal):
+        return _format_dec(number, decimal_sep, decimal_pos, thousand_sep, grouping)
 
-        if decimal_pos is not None:
-            # If the provided number is too small to affect any of the visible
-            # decimal places, consider it equal to '0'.
-            cutoff = Decimal('0.' + '1'.rjust(decimal_pos, '0'))
-            if abs(number) < cutoff:
-                number = Decimal('0')
-
-        # Format values with more than 200 digits (an arbitrary cutoff) using
-        # scientific notation to avoid high memory usage in {:f}'.format().
-        _, digits, exponent = number.as_tuple()
-        if abs(exponent) + len(digits) > 200:
-            number = '{:e}'.format(number)
-            coefficient, exponent = number.split('e')
-            # Format the coefficient.
-            coefficient = format(
-                coefficient, decimal_sep, decimal_pos, grouping,
-                thousand_sep, force_grouping, use_l10n,
-            )
-            return '{}e{}'.format(coefficient, exponent)
-        else:
-            str_number = '{:f}'.format(number)
+    # Don't unnecessarily convert number to string for speed
+    if isinstance(number, str):
+        return _format_string(number, decimal_sep, decimal_pos, thousand_sep, grouping)
+
+    str_number = str(number)
+    # Treat potentially very large/small floats as Decimals.
+    if isinstance(number, float) and 'e' in str_number.lower():
+        return _format_dec(Decimal(str_number), decimal_sep, decimal_pos, thousand_sep, grouping)
+
+    # Format all other cases as a string
+    return _format_string(str_number, decimal_sep, decimal_pos, thousand_sep, grouping)
+
+
+def _format_int(number, decimal_sep, decimal_pos, thousand_sep, grouping):
+    """
+    Format an integer as defined by the arguments, returning as a string.
+
+    This function only handles when grouping is 0 or 3.
+    """
+    if grouping and (number > 999 or number < -999):
+        # f'{number:,}' returns a string with a ',' as a thousand separator
+        number = f'{number:,}'
+        if thousand_sep != ',':
+            number = number.replace(',', thousand_sep)
     else:
-        str_number = str(number)
+        number = str(number)
+    if decimal_pos:
+        return number + decimal_sep + '0' * decimal_pos
```

Comment:
Why don't you do the decimal precision as part of the format expression? Something like:
```python
number = ("{:_.%sf}" % decimal_pos).format(number)
```

Reply:
Using `format` to take care of the decimals here is alot slower. It's slower for both large and small, but for large numbers can take 2 or 3 times as long. (Which is a shame, it looks much nicer). So I've left it as it is.

Reply:
Or you can still use f-strings for this too:
```python
number = f'{number:_.{decimal_pos}f}'
```

Reply:
@pope1ni This one seems to be slower as well. Especially as you then have to replace the decimal point when `decimal_sep` is something other than a full-stop/period.